cpp/src/traversal/extract_bfs_paths_impl.cuh

```diff
@@ -0,0 +1,378 @@
+                      current_position.end(),
+                      current_position.data(),
+                      detail::decrement_position{});
+
+    if constexpr (multi_gpu) {
+      rmm::device_uvector<vertex_t> tmp_frontier(current_frontier.size(), handle.get_stream());
+      rmm::device_uvector<vertex_t> original_position(current_frontier.size(), handle.get_stream());
+      rmm::device_uvector<int> original_rank(current_frontier.size(), handle.get_stream());
+
+      raft::copy(tmp_frontier.data(),
+                 current_frontier.begin(),
+                 current_frontier.size(),
+                 handle.get_stream());
+      detail::sequence_fill(
+        handle.get_stream(), original_position.data(), original_position.size(), vertex_t{0});
+      detail::fill(original_rank.size(), original_rank.data(), handle.get_comms().get_rank());
+
+      auto tuple_begin = thrust::make_zip_iterator(
+        thrust::make_tuple(tmp_frontier.begin(), original_position.begin(), original_rank.begin()));
+
+      std::forward_as_tuple(std::tie(tmp_frontier, original_position, original_rank), std::ignore) =
+        groupby_gpuid_and_shuffle_values(
+          handle.get_comms(),
+          tuple_begin,
+          tuple_begin + tmp_frontier.size(),
+          [key_func =
+             detail::mg_partition_vertices<vertex_t>{
+               handle.get_comms().get_rank(),
+               d_vertex_partition_offsets.begin(),
+               d_vertex_partition_offsets.end()}] __device__(auto val) {
+            return key_func(thrust::get<0>(val));
+          },
+          handle.get_stream());
+
+      tuple_begin = thrust::make_zip_iterator(
+        thrust::make_tuple(tmp_frontier.begin(), original_position.begin(), original_rank.begin()));
+
+      thrust::transform(
+        handle.get_thrust_policy(),
+        tuple_begin,
+        tuple_begin + tmp_frontier.size(),
+        tuple_begin,
+        detail::mg_lookup_predecessor<vertex_t>{graph_view.get_local_vertex_first(), predecessors});
+
+      std::forward_as_tuple(std::tie(tmp_frontier, original_position, original_rank), std::ignore) =
+        groupby_gpuid_and_shuffle_values(
+          handle.get_comms(),
+          tuple_begin,
+          tuple_begin + tmp_frontier.size(),
+          [] __device__(auto val) { return thrust::get<2>(val); },
+          handle.get_stream());
+
+      thrust::for_each_n(handle.get_thrust_policy(),
+                         thrust::make_zip_iterator(
+                           thrust::make_tuple(tmp_frontier.begin(), original_position.begin())),
+                         tmp_frontier.size(),
+                         [d_current_frontier = current_frontier.data()] __device__(auto tuple) {
+                           auto v        = thrust::get<0>(tuple);
+                           auto position = thrust::get<1>(tuple);
+
+                           d_current_frontier[position] = v;
+                         });
```

Comment:
So, if I am not mistaken, this collects predecessors from the distributed predecessors array. And if we implement collect_values_for_vertices (as I mentioned in the code collecting remote distances from the distributed distances array), I think this code can be largely simplified.

Reply:
Yes.  Do you think this simplification should be done in this PR, or can we defer it to the MG PR which will be in 21.12?

I could strip out all of the MG code from this PR and make the suggested improvements in the MG PR.

Reply:
Doesn't matter for me. You can either add a FIXME, strip out the code, or make fix in this release, or whatever... I just want to make sure we eventually make this code simpler & more reusable.